src/main/host/memory_manager.rs

```diff
@@ -0,0 +1,769 @@
+    // needed.
+    fn post_exec_cleanup_if_needed(&self, thread: &impl Thread) {
+        if !self.need_post_exec_cleanup.get() {
+            return;
+        }
+        self.reset_and_unmap_all();
+        let shm_file = self.shm_file.borrow();
+        *self.regions.borrow_mut() = get_regions(thread.get_system_pid());
+        *self.heap.borrow_mut() = get_heap(&*shm_file, thread, &mut *self.regions.borrow_mut());
+        let stack_end = map_stack(&*shm_file, &mut *self.regions.borrow_mut());
+        *self.stack_copied.borrow_mut() = stack_end..stack_end;
+        self.need_post_exec_cleanup.set(false);
+    }
+
+    /// Gets a reference to an object in the plugin's memory.
+    /// SAFETY
+    /// * The pointer must point to readable value of type T.
+    /// * Returned ref mustn't be accessed after Thread runs again or flush is called.
+    #[allow(dead_code)]
+    pub unsafe fn _get_ref<T>(&self, thread: &impl Thread, src: PluginPtr) -> Result<&T, i32> {
+        let raw = self.get_readable_ptr(thread, src, std::mem::size_of::<T>())?;
+        Ok(&*(raw as *const T))
+    }
+
+    /// Gets a mutable reference to an object in the plugin's memory.
+    /// SAFETY
+    /// * The pointer must point to a writeable value of type T.
+    /// * Returned ref mustn't be accessed after Thread runs again or flush is called.
+    #[allow(dead_code)]
+    pub unsafe fn get_mut_ref<T>(
+        &self,
+        thread: &impl Thread,
+        src: PluginPtr,
+    ) -> Result<&mut T, i32> {
+        let raw = self.get_writeable_ptr(thread, src, std::mem::size_of::<T>())?;
+        Ok(&mut *(raw as *mut T))
+    }
+
+    /// Gets a slice of the plugin's memory.
+    /// SAFETY
+    /// * The pointer must point to a readable array of type T and at least size `len`.
+    /// * Returned slice mustn't be accessed after Thread runs again or flush is called.
+    #[allow(dead_code)]
+    pub unsafe fn get_slice<T>(
+        &self,
+        thread: &impl Thread,
+        src: PluginPtr,
+        len: usize,
+    ) -> Result<&[T], i32> {
+        let raw = self.get_readable_ptr(thread, src, std::mem::size_of::<T>() * len)?;
+        Ok(std::slice::from_raw_parts(raw as *const T, len))
+    }
+
+    /// Gets a mutable slice of the plugin's memory.
+    /// SAFETY
+    /// * The pointer must point to a writeable array of type T and at least size `len`.
+    /// * Returned slice mustn't be accessed after Thread runs again or flush is called.
+    #[allow(dead_code)]
+    pub unsafe fn get_mut_slice<T>(
+        &self,
+        thread: &impl Thread,
+        src: PluginPtr,
+        len: usize,
+    ) -> Result<&mut [T], i32> {
+        let raw = self.get_writeable_ptr(thread, src, std::mem::size_of::<T>() * len)?;
+        Ok(std::slice::from_raw_parts_mut(raw as *mut T, len))
+    }
+
+    /// We take a mutable reference here to ensure there are no outstanding borrowed
+    /// references/pointers, which might otherwise become invalidated if shadow's mapping of the
```

Comment:
How does taking a mutable reference ensure there are no borrowed references/pointers?

Reply:
> Rust won't allow you to call a mutable method of an object from which there are any borrowed lifetimes. e.g.
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=368f25542a78ace3935d00214dd260eb

I'm still not convinced here. The `get_mapped_ptr()` returns a `*const c_void` which in `get_ref()` is dereferenced and converted to a Rust reference with `unsafe { &*(raw as *const T) }`. I think this operation will cause the `Ref` to go out of scope, so `RefCell` will think it's no longer borrowed? So I think Rust will allow you to call that method with a mutable reference to the memory manager, even if there are outstanding mutable or immutable references. https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=cbb4bf1fef3e098a01b236df26eb15fd

Reply:
Yeah, we're counting on the static safety in this case, not the Ref (even before removing it). The reference returned e.g. by MemoryManager::get_ref is statically borrowed from the MemoryManager, so calling any other mutable methods on MemoryManager while that borrowed reference is alive would be a compile-time error.

As you show in your example, that static check doesn't work for pointers, only for refs. It works when calling e.g. MemoryManager::get_ref, but the signature is implicitly `pub unsafe fn get_ref<'a, T>(&'a mut self, thread: &'a impl Thread, src: PluginPtr) -> Result<&'a T, i32>` - from the caller's point of view the ref is borrowed from the memorymanager and the thread, even though internally the ref is constructed from a pointer.

If the caller uses `MemoryManager::get_readable_ptr` to get a pointer and converts it to a reference itself, as in your example, all bets are off again, since the reference won't be viewed as borrowed from the MemoryManager.

Reply:
Ah okay this clears it up for me, thanks!